File: ios/XPush/ChannelCore.swift
//
//  ChannelCore.swift
//  SampleApp
//
//  Created by James Jung on 2016. 5. 29..
//  Copyright © 2016년 Facebook. All rights reserved.
//

import Foundation
//
//  ChannelCore.swift
//
//
//  Created by James Jung on 2016. 5. 29..
//
//

import Foundation

class ChannelCore: NSObject {
  var socket: SocketIOClient!

  var mAppId: String?;
  var mUserId: String?;
  var mDeviceId: String?;

  var mChannelId: String?;
  var mServerUrl: String?;
  var mServerName: String?;
  var connectionSuccessCallback: RCTResponseSenderBlock!;

  var mEvents: [String:NormalCallback]!;

  /// Type safe way to create a new SocketIOClient. opts can be omitted
  internal init(mAppId: String, mUserId: String, mDeviceId: String, mChannelId: String, mServerUrl: String, mServerName: String) {
    self.mAppId = mAppId;
    self.mUserId = mUserId;
    self.mDeviceId = mDeviceId;
    self.mChannelId = mChannelId;
    self.mServerUrl = mServerUrl;
    self.mServerName = mServerName;
  };

  internal func connect(handlers: [String:NormalCallback]){
    let url = self.mServerUrl!;

    var params:[String:AnyObject] = [String:AnyObject]();
    params["A"] = self.mAppId as AnyObject?;
    params["C"] = self.mChannelId as AnyObject?;
    params["S"] = self.mServerName as AnyObject?;
    params["D"] = self.mDeviceId as AnyObject?;
    params["U"] = self.mUserId as AnyObject?;

    socket = SocketIOClient(socketURL: URL(string:url)!, config:[.log(true), .forceNew(true), .connectParams(params), .nsp("/channel") ]);

    mEvents = handlers;

    for handler in handlers {
      socket?.on(handler.0, callback: handler.1);
    }

    self.socket.connect();
  };

  internal func sendText(message:String){
    if self.socket != nil {
      var json:[String:AnyObject] = [String:AnyObject]();
      var data:[String:AnyObject] = [String:AnyObject]();
      var user:[String:AnyObject] = [String:AnyObject]();

      user["U"] = self.mUserId as AnyObject?;
      data["UO"] = user as AnyObject?;
      data["MG"] = message as AnyObject?;

      json["DT"] = data as AnyObject?;
      json["NM"] = "message" as AnyObject?;

      self.socket.emit("send", json );
    }
  };

  internal func sendData(param:[String:AnyObject]){
    if self.socket != nil {
      var json:[String:AnyObject] = [String:AnyObject]();
      var user:[String:AnyObject] = [String:AnyObject]();
      var data = param;

      user["U"] = self.mUserId as AnyObject?;
      data["UO"] = user as AnyObject?;

      json["DT"] = data as AnyObject?;
      json["NM"] = "message" as AnyObject?;

      self.socket.emit("send", json );
    }
  };

  internal func disconnect(){
    if self.socket != nil {

      for handler in self.mEvents {
        self.off(event: handler.0);
      }

      self.socket.disconnect();
    }
  };

  internal func off(event:String) {
    self.socket.off(event);
  };

  internal func channelGet(callback:@escaping ([String:AnyObject]) -> Void){

    self.socket.emitWithAck("channel.get").timingOut(after: 0) {data in
      if let res = data[0] as? [String:AnyObject] {
        callback(res);
      }
    }

  }

  internal func channelJoin(users:NSArray, callback:@escaping ([String:AnyObject]) -> Void){
    var json:[String:AnyObject] = [String:AnyObject]();
    //json["U"] = param.valueForKey("users");
    json["U"] = users;

    self.socket.emitWithAck("channel.join").timingOut(after: 0) {data in
      if let res = data[0] as? [String:AnyObject] {
        callback(res);
      }
    }
  }

  internal func banFromChannel(users:NSArray, callback:@escaping ([String:AnyObject]) -> Void){
    var json:[String:AnyObject] = [String:AnyObject]();
    json["U"] = users;

    self.socket.emitWithAck("channel.ban").timingOut(after: 0) {data in
      if let res = data[0] as? [String:AnyObject] {
        callback(res);
      }
    }
  }

  internal func channelLeave(callback:@escaping ([String:AnyObject]) -> Void){

    self.socket.emitWithAck("channel.leave").timingOut(after: 0) {data in
      if let res = data[0] as? [String:AnyObject] {
        callback(res);
      }
    }
  }
}

File: android/src/main/java/io/xpush/core/ChannelCore.java
package io.xpush.core;

import android.text.TextUtils;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.HashMap;

import io.socket.client.Ack;
import io.socket.client.IO;
import io.socket.client.Socket;
import io.socket.emitter.Emitter;

public class ChannelCore {

    private static final String TAG = ChannelCore.class.getSimpleName();

    private String mChannelId;
    private String mServerUrl;
    private String mServerName;

    private String mUserId;
    private String mAppId;
    private String mDeviceId;

    private Socket mChannelSocket;

    private HashMap<String, Emitter.Listener> mEvents;

    public ChannelCore() {
    }

    public ChannelCore(String appId, String mUserId, String mDeviceId, String channelId, String serveUrl, String serverName) {
        this();
        this.mAppId = appId;
        this.mDeviceId = mDeviceId;
        this.mUserId = mUserId;
        this.mChannelId = channelId;
        this.mServerUrl = serveUrl;
        this.mServerName = serverName;
    }

    public void connect(HashMap<String, Emitter.Listener> events) {

        mEvents = events;
        String url = mServerUrl + "/channel";

        IO.Options opts = new IO.Options();
        opts.forceNew = true;

        opts.query = "A=" + mAppId + "&C=" + mChannelId + "&S=" + mServerName + "&D=" + mDeviceId + "&U=" + mUserId;

        mChannelSocket = null;

        try {
            mChannelSocket = IO.socket(url, opts);
        } catch (URISyntaxException e) {
            throw new RuntimeException(e);
        }

        if (events != null) {
            for (String eventName : events.keySet()) {
                this.on(eventName, events.get(eventName));
            }
        }

        mChannelSocket.connect();
    }

    public void sendText(String message) {
        sendMessage(message, null, null, null);
    }

    public void sendText(String message, String type) {
        sendMessage(message, type, null, null);
    }

    public void sendImage(String message, int width, int height){
        JSONObject metaData = new JSONObject();

        try {

            metaData.put("W", width);
            metaData.put("H", height);

        } catch (JSONException e) {
            e.printStackTrace();
        }
        sendMessage(message, "IM", metaData, null );
    }

    public void sendText(String message, String type, ArrayList<String> users) {
        sendMessage(message, type, null, users);
    }

    public void sendText(String message, String type, JSONObject metaData, ArrayList<String> users) {

        JSONObject json = new JSONObject();
        JSONObject data = new JSONObject();
        JSONObject user = new JSONObject();

        try {

            //user.put("I", xpushSession.getImage());
            user.put("U", mUserId);
            //user.put("NM", xpushSession.getName());

            data.put("UO", user);
            data.put("MG", message);
            if( type != null ) {
                data.put("TP", type);
            }

            if( users != null ) {
                data.put("US", TextUtils.join("@!@", users));
            }

            if( metaData != null ){
                data.put("MD", metaData);
            }

            json.put("DT", data);
            json.put("NM", "message");

        } catch (JSONException e) {
            e.printStackTrace();
        }

        if (mChannelSocket != null && mChannelSocket.connected()) {
            mChannelSocket.emit("send", json);
        }
    }

    public void sendData(JSONObject data) {

        JSONObject json = new JSONObject();
        JSONObject user = new JSONObject();

        try {
            user.put("U", mUserId);
            data.put("UO", user);

            json.put("DT", data);
            json.put("NM", "message");

        } catch (JSONException e) {
            e.printStackTrace();
        }

        if (mChannelSocket != null && mChannelSocket.connected()) {
            mChannelSocket.emit("send", json);
        }
    }

    public void emit(final String event, final Object... args) {
        mChannelSocket.emit(event, args);
    }

    public void on(String event, Emitter.Listener fn) {
        mChannelSocket.on(event, fn);
    }

    public void off(String event) {
        mChannelSocket.off(event);
    }

    public void off(String event, Emitter.Listener fn) {
        mChannelSocket.off(event, fn);
    }

    public boolean connected() {
        if (mChannelSocket == null) {
            return false;
        }

        return mChannelSocket.connected();
    }

    public void disconnect() {
        if (mChannelSocket != null) {
            for (String eventName : mEvents.keySet()) {
                this.off(eventName);
            }
            Log.d(TAG, "=== disconnect ===");
            mChannelSocket.disconnect();
        }
    }

    public void getMessageUnread(long lastReceiveTime, final CallbackEvent callback) {
        if (mChannelSocket != null) {

            JSONObject jsonObject = new JSONObject();
            try {

                if (lastReceiveTime > 0) {
                    jsonObject.put("TS", lastReceiveTime);
                }

            } catch (JSONException e) {
                e.printStackTrace();
            }

            mChannelSocket.emit("message.unread", jsonObject, new Ack() {
                @Override
                public void call(Object... args) {
                    JSONObject response = (JSONObject) args[0];
                    if (response.has("status")) {
                        try {
                            if ("ok".equalsIgnoreCase(response.getString("status"))) {
                                JSONArray messages = (JSONArray) response.get("result");
                                callback.call(messages);
                            }
                        } catch (JSONException e) {
                            e.printStackTrace();
                            callback.call(new JSONArray());
                        }
                    } else {
                        callback.call(new JSONArray());
                    }
                }
            });
        }
    }


    public void channelJoin(ArrayList<String> userIdArrayList, final CallbackEvent callback) {

        JSONArray userArray = new JSONArray();
        for( String userId : userIdArrayList ){
            userArray.put(userId);
        }

        JSONObject data = new JSONObject();
        try {
            data.put( "U", userArray );
        } catch (JSONException e) {
            e.printStackTrace();
        }

        mChannelSocket.emit("channel.join", data, new Ack() {
            @Override
            public void call(Object... args) {
                JSONObject response = (JSONObject) args[0];
                callback.call(response);
            }
        });
    }

    public void banFromChannel(ArrayList<String> userIdArrayList, final CallbackEvent callback) {

        JSONArray userArray = new JSONArray();
        for( String userId : userIdArrayList ){
            userArray.put(userId);
        }

        JSONObject data = new JSONObject();
        try {
            data.put( "U", userArray );
        } catch (JSONException e) {
            e.printStackTrace();
        }

        mChannelSocket.emit("channel.ban", data, new Ack() {
            @Override
            public void call(Object... args) {
                JSONObject response = (JSONObject) args[0];
                callback.call(response);
            }
        });
    }

    public void channelGet(final CallbackEvent callback) {

        mChannelSocket.emit("channel.get", new Ack() {
            @Override
            public void call(Object... args) {
                JSONObject response = (JSONObject) args[0];
                callback.call(response);
            }
        });
    }

    public void channelLeave(final CallbackEvent callback) {

        mChannelSocket.emit("channel.leave", new Ack() {
            @Override
            public void call(Object... args) {
                JSONObject response = (JSONObject) args[0];
                callback.call(response);
            }
        });
    }
}